django/template/library.py

```diff
@@ -153,6 +153,86 @@ def compile_func(parser, token):
         else:
             raise ValueError("Invalid arguments provided to simple_tag")
 
+    def simple_block_tag(self, func=None, takes_context=None, name=None):
```

Comment:
I can't help noticing the amount of duplicated code with `simple_tag`. I tried to give it a go at refactoring this, my solution still has an issue somewhere which I can't find in the time frame I have, but I would love if you could give it a go and share your thoughts. Thanks!

Reply:
I suspect reducing duplication is going to require the changes deeper down the stack, making them more brittle and higher impact. But with a fresh pair of eyes I'll give them a go and see where we get! It might be possible abstract them out at the very least.

Reply:
> I suspect reducing duplication is going to require the changes deeper down the stack, making them more brittle and higher impact.

Yeah, maybe...

> But with a fresh pair of eyes I'll give them a go and see where we get! It might be possible abstract them out at the very least.

Perhaps take a "quick" look and let me know what you find. I gave this a "quick" try yesterday and I ended up in a 2 hours rabbit hole. What I tried is (but is not working, when running the tests, `self.func` is unexpectedly `None`:

```python
    def _build_tag(
        self,
        build_node,
        node_name,
        func=None,
        takes_context=None,
        name=None,
        validate_params=None,
    ):
        """Register a callable as a compiled template tag."""

        def dec(func):
            (
                params,
                varargs,
                varkw,
                defaults,
                kwonly,
                kwonly_defaults,
                _,
            ) = getfullargspec(unwrap(func))
            function_name = name or func.__name__

            @wraps(func)
            def compile_func(parser, token):
                if validate_params is not None:
                    validated_params = validate_params(function_name, params)
                else:
                    validated_params = params
                bits = token.split_contents()[1:]
                target_var = None
                if len(bits) >= 2 and bits[-2] == "as":
                    target_var = bits[-1]
                    bits = bits[:-2]
                return build_node(
                    parser,
                    bits,
                    validated_params,
                    varargs,
                    varkw,
                    defaults,
                    kwonly,
                    kwonly_defaults,
                    function_name=function_name,
                    target_var=target_var,
                )

            self.tag(function_name, compile_func)
            return func

        if func is None:
            # @register.<node_name>(...)
            return dec
        elif callable(func):
            # @register.<node_name>
            return dec(func)
        else:
            raise ValueError(f"Invalid arguments provided to {node_name}")


    def simple_tag(self, func=None, takes_context=None, name=None):
        """
        Register a callable as a compiled template tag. Example:

        @register.simple_tag
        def hello(*args, **kwargs):
            return 'world'
        """

        def build_node(*args, function_name, target_var):
            args, kwargs = parse_bits(
                *args,
                takes_context=takes_context,
                name=function_name,
            )
            return SimpleNode(func, takes_context, args, kwargs, target_var)

        return self._build_tag(build_node, "simple_tag", func, takes_context, name)

    def simple_block_tag(self, func=None, takes_context=None, name=None):
        """
        Register a callable as a compiled block template tag. Example:

        @register.simple_block_tag
        def hello(content):
            return 'world'
        """

        def validate_params(function_name, params):
            tag_params = params.copy()
            if takes_context:
                if len(tag_params) >= 2 and tag_params[1] == "content":
                    del tag_params[1]
                else:
                    raise TemplateSyntaxError(
                        f"'{function_name}' is decorated with takes_context=True so"
                        " it must have a first argument of 'context' and a second "
                        "argument of 'content'"
                    )
            else:
                if tag_params and tag_params[0] == "content":
                    del tag_params[0]
                else:
                    raise TemplateSyntaxError(
                        f"'{function_name}' must have a first argument of 'content'"
                    )
            return tag_params

        def build_node(parser, *args, function_name, target_var):
            nodelist = parser.parse((f"end{function_name}",))
            parser.delete_first_token()
            args, kwargs = parse_bits(
                parser,
                *args,
                takes_context=takes_context,
                name=function_name,
            )
            return SimpleBlockNode(
                nodelist, func, takes_context, args, kwargs, target_var
            )

        return self._build_tag(
            build_node, "simple_block_tag", func, takes_context, name, validate_params
        )
```

Reply:
Whilst there is duplication, I'm not sure how valuable it is splitting it out. Other than the small skeleton of decorating a function and defining it as a tag, the only duplicated parts are:

- Handling a trailing "as"
- Calling `parse_bits`
- Returning a `Node` instance

Trying to remove duplication by separating them out will just make things harder to follow and maintain.

I'd strongly suggest we leave it as-is, and potentially improve it later if a third similar-looking tag is defined (which it may when I write an `include_block` tag).

Reply:
Fair enough! I agree with your reasoning. :100: 

I'm out of bandwidth this week to complete the review but I'll do so early next week.